### PR TITLE
Remove h7 and h8 headings

### PR DESCRIPTION
**Pull request by wbamberg**
Fixes https://github.com/matrix-org/matrix-doc/issues/2898.

The new spec doesn't support "h7" or "h8" because these are not HTML elements. There are a few places in the spec that want to use this level of heading:

```
./content/client-server-api/modules/secrets.md:####### `m.pbkdf2`
./content/client-server-api/modules/instant_messaging.md:####### Fallbacks and event representation
./content/client-server-api/modules/instant_messaging.md:######## Stripping the fallback
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.text`, `m.notice`, and unrecognised message types
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.emote`
./content/client-server-api/modules/instant_messaging.md:######## Fallback for `m.image`, `m.video`, `m.audio`, and `m.file`
./content/client-server-api/modules/push.md:####### Tweaks
./content/client-server-api/modules/push.md:####### Default Override Rules
./content/client-server-api/modules/push.md:######## `.m.rule.master`
./content/client-server-api/modules/push.md:######## `.m.rule.suppress_notices`
./content/client-server-api/modules/push.md:######## `.m.rule.invite_for_me`
./content/client-server-api/modules/push.md:######## `.m.rule.member_event`
./content/client-server-api/modules/push.md:######## `.m.rule.contains_display_name`
./content/client-server-api/modules/push.md:######## `.m.rule.tombstone`
./content/client-server-api/modules/push.md:######## `.m.rule.roomnotif`
./content/client-server-api/modules/push.md:####### Default Content Rules
./content/client-server-api/modules/push.md:######## `.m.rule.contains_user_name`
./content/client-server-api/modules/push.md:####### Default Underride Rules
./content/client-server-api/modules/push.md:######## `.m.rule.call`
./content/client-server-api/modules/push.md:######## `.m.rule.encrypted_room_one_to_one`
./content/client-server-api/modules/push.md:######## `.m.rule.room_one_to_one`
./content/client-server-api/modules/push.md:######## `.m.rule.message`
./content/client-server-api/modules/push.md:######## `.m.rule.encrypted`
```

There are basically two options here.

1. Define extra heading levels using something like https://www.w3.org/TR/WCAG20-TECHS/ARIA12.html#ARIA12-ex2
2. Update the content so it doesn't nest headings this deeply.

I'd like us to avoid (1) if possible, because:

* it's more complicated to implement: there will be several different places to implement custom behaviour (ToC, auto-numbering, anchor and anchor-link icon) it may not work well in some edge cases and there will be problems with accessibility (https://css-tricks.com/how-to-section-your-html/#what-happens-when-you-need-h7). This will also make the spec platform harder to maintain.

* I think having deep heading levels is also tough on readers.

So this PR attempts option (2). But I can't pretend this is ideal either. It's not surprising that all these instances are in client-server modules: in the client-server spec all modules are H3 at the top level, so you really only have 3 levels to play with inside the module document itself. Still I'd like to present this as my preferred option, and if it doesn't look reasonable, we'll look at option (1).

******

## instant_messaging.md

Here I've folded  `"H6 Rich replies"` into `"H5 Forming relationships between events"`: since rich replies are currently the only form of events that reference other events, this seemed reasonable. I then pushed up `"H7 Fallbacks and event representation"` to `"H5 Fallbacks for rich replies"`, which means we can keep the subsections for handling fallbacks under there as H6 headings. I think it's not ideal that "Fallbacks for rich replies" is not under "Rich replies" though.

## push.md

First, I removed the separate `"H7 Tweaks"` and folded that content directly under the definition for `set_tweak`. This seems to make sense.

Then I've pushed up `"H6 Predefined Rules"` to `"H5 Predefined Rules"`, so it's alongside `"H5 Push Rules"` instead of underneath it. I don't really think this is ideal, but it lets me keep `"H6 Default Override Rules"` and so on underneath `"Predefined Rules"`. Then under there I've just made the rule names bold, and it seems quite readable at least.

## secrets.md

Here I just removed the heading, and it seems fine like that.

### CODE REVIEW

File: content/client-server-api/modules/secrets.md [hunk 1]
```diff
@@ -222,8 +222,6 @@ has an `algorithm` property that indicates how to generate the key from
 the passphrase. Other properties of the `passphrase` property are
 defined by the `algorithm` specified.
 
-####### `m.pbkdf2`
-
 For the `m.pbkdf2` algorithm, the `passphrase` property has the
```
Comment (uhoreg) [inline content/client-server-api/modules/secrets.md:225]: The intention is that there can be different algorithms defined, although only one is defined currently.  So maybe we should prefix the paragraph with a short explanation, maybe something along the lines of:

```suggestion
Currently, the only algorithm defined is `m.pbkdf2`. For the `m.pbkdf2` algorithm, the `passphrase` property has the
```

**Review by anoadragon453 — APPROVED** [review]
While indeed not the most ideal situation, I agree that it's the best course of action for right now. It would be unfortunate to have to go and redesign everything at this point. It is, however, something we'll likely need to do in the future, as I'm not sure things will just happen to work out this way a year down the line.

We should put the detail that headings should not be greater than h6 somewhere where contributors can see it. You could make that part of your overall documentation PR perhaps?

**Review by KitsuneRal — APPROVED** [review]
Looks like reasonable flattening.

**Comment by wbamberg** [comment]
> We should put the detail that headings should not be greater than h6 somewhere where contributors can see it. You could make that part of your overall documentation PR perhaps?

Yes, I will make sure this is part of the meta-docs.

**Pull request merged**